File: scripts/dev_tools/codex_native_converter/inventory.py
"""Discover and normalize source artifacts for the Codex-native converter.

Purpose:
    Provide deterministic source-file enumeration for supported ecosystems while
    preventing caller-selected paths from escaping the declared source root.

Usage:
    The converter engine calls ``normalize_selected_paths`` first when the user
    provides an explicit subset, then calls ``discover_source_artifacts`` to
    collect the final artifact list.

Flow:
    Supported top-level source surfaces are selected by ecosystem, candidate
    files are enumerated beneath those surfaces, and the result is sorted by
    normalized relative path.

Invariants / Constraints:
    All returned paths are source-root-relative ``Path`` values using a stable
    path ordering that is independent of host operating system traversal order.

Side Effects:
    Reads the filesystem metadata under the declared source root.
"""

from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

from scripts.dev_tools.codex_native_converter.models import SourceEcosystem

if TYPE_CHECKING:
    from collections.abc import Iterable

_SUPPORTED_ROOTS: dict[SourceEcosystem, tuple[Path, ...]] = {
    SourceEcosystem.GITHUB_COPILOT: (
        Path(".github/copilot-instructions.md"),
        Path(".github/instructions"),
        Path(".github/skills"),
        Path(".github/agents"),
        Path(".github/prompts"),
    ),
    SourceEcosystem.CLAUDE: (
        Path("CLAUDE.md"),
        Path(".claude/skills"),
        Path(".claude/agents"),
        Path(".claude/hooks"),
        Path(".claude/settings.json"),
        Path(".claude/rules"),
    ),
}


def _normalize_relative_path(source_root: Path, candidate_path: Path) -> Path:
    """Normalize one candidate path relative to the declared source root.

    Purpose:
        Convert an absolute or relative caller-provided path into a validated
        source-root-relative path.

    Args:
        source_root (Path): Absolute or relative root that bounds the converter
            input set.
        candidate_path (Path): Caller-provided source path to validate.

    Returns:
        Path: A normalized relative path beneath ``source_root``.

    Raises:
        ValueError: Raised when the candidate path escapes the declared source
            root.

    Side Effects:
        Resolves filesystem paths to normalize parent-segment references.
    """

    resolved_root = source_root.resolve()
    resolved_candidate = (
        candidate_path.resolve()
        if candidate_path.is_absolute()
        else (resolved_root / candidate_path).resolve()
    )
    try:
        return resolved_candidate.relative_to(resolved_root)
    except ValueError as error:
        raise ValueError(
            f"Selected path escapes the declared source root: {candidate_path}"
        ) from error


def normalize_selected_paths(
    source_root: Path,
    selected_paths: Iterable[Path],
) -> tuple[Path, ...]:
    """Normalize selected source paths beneath the declared source root.

    Purpose:
        Validate user-selected paths and return them in deterministic normalized
        relative-path order.

    Args:
        source_root (Path): Root directory that bounds valid source input.
        selected_paths (Iterable[Path]): Caller-selected paths beneath the root.

    Returns:
        tuple[Path, ...]: Unique normalized relative paths sorted by POSIX text.

    Raises:
        ValueError: Raised when any selected path escapes the declared source
            root.

    Side Effects:
        Resolves filesystem paths while validating the selected paths.
    """

    normalized_paths = {
        _normalize_relative_path(source_root, candidate_path)
        for candidate_path in selected_paths
    }
    return tuple(sorted(normalized_paths, key=lambda path: path.as_posix()))


def _iter_supported_artifacts(
    source_root: Path,
    source_ecosystem: SourceEcosystem,
) -> tuple[Path, ...]:
    """Yield supported artifacts for one source ecosystem.

    Purpose:
        Enumerate existing supported files below the configured ecosystem roots.

    Args:
        source_root (Path): Source tree root to scan.
        source_ecosystem (SourceEcosystem): Ecosystem that defines supported
            surfaces.

    Returns:
        tuple[Path, ...]: Unique supported relative file paths beneath the
            source root.

    Raises:
        None.

    Side Effects:
        Reads directory contents below the supported roots.
    """

    discovered_paths: set[Path] = set()
    resolved_root = source_root.resolve()

    # Enumerate only the supported ecosystem surfaces so classification starts
    # from the approved v1 input set.
    for supported_root in _SUPPORTED_ROOTS[source_ecosystem]:
        absolute_supported_root = (resolved_root / supported_root).resolve()
        if not absolute_supported_root.exists():
            continue
        if absolute_supported_root.is_file():
            discovered_paths.add(absolute_supported_root.relative_to(resolved_root))
            continue

        # Walk supported directories recursively while preserving files only.
        for candidate_path in absolute_supported_root.rglob("*"):
            if candidate_path.is_file():
                discovered_paths.add(candidate_path.relative_to(resolved_root))

    return tuple(sorted(discovered_paths, key=lambda path: path.as_posix()))


def discover_source_artifacts(
    source_root: Path,
    source_ecosystem: SourceEcosystem,
    selected_paths: Iterable[Path] | None = None,
) -> tuple[Path, ...]:
    """Discover source artifacts in deterministic normalized-relative-path order.

    Purpose:
        Produce the set of source files that the converter should classify for a
        given run.

    Args:
        source_root (Path): Root directory of the source ecosystem tree.
        source_ecosystem (SourceEcosystem): Declared source ecosystem.
        selected_paths (Iterable[Path] | None): Optional caller-selected subset
            of files or directories beneath the source root.

    Returns:
        tuple[Path, ...]: Supported source artifact paths relative to the source
            root, sorted by POSIX path text.

    Raises:
        ValueError: Raised when selected paths escape the source root.

    Side Effects:
        Reads directory contents beneath the source root.
    """

    all_artifacts = _iter_supported_artifacts(source_root, source_ecosystem)
    if selected_paths is None:
        return all_artifacts

    normalized_selected_paths = normalize_selected_paths(source_root, selected_paths)
    if not normalized_selected_paths:
        return all_artifacts

    selected_set = set(normalized_selected_paths)
    matched_artifacts: list[Path] = []

    # Include files selected directly and files contained beneath selected
    # directories so the caller can target either shape deterministically.
    for artifact_path in all_artifacts:
        if artifact_path in selected_set:
            matched_artifacts.append(artifact_path)
            continue
        if any(parent in artifact_path.parents for parent in selected_set):
            matched_artifacts.append(artifact_path)

    return tuple(sorted(matched_artifacts, key=lambda path: path.as_posix()))

File: docs/features/active/2026-08-17-orchestrator-remediation-loop-control-484/evidence/qa-gates/typescript-benchmark-build/lib/codex-native-converter/inventory.js
"use strict";
/**
 * Discover and normalize source artifacts for the Codex-native converter.
 *
 * Purpose:
 *     Provide deterministic source-file enumeration for supported ecosystems
 *     while preventing caller-selected paths from escaping the declared source
 *     root. Ported from `inventory.py`; filesystem access flows through the
 *     injected {@link FileSystem} rather than `node:fs`.
 *
 * Flow:
 *     Supported top-level source surfaces are selected by ecosystem, candidate
 *     files are enumerated beneath those surfaces, and the result is sorted by
 *     normalized relative path.
 *
 * Invariants:
 *     All returned paths are source-root-relative POSIX strings using a stable
 *     ordering independent of host traversal order.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeSelectedPaths = normalizeSelectedPaths;
exports.discoverSourceArtifacts = discoverSourceArtifacts;
const file_system_1 = require("../file-system");
const models_1 = require("./models");
/**
 * Supported discovery roots per ecosystem.
 *
 * Mirrors the Python `_SUPPORTED_ROOTS` mapping. Each entry is a source-root
 * relative POSIX path that may be a file or a directory.
 */
const SUPPORTED_ROOTS = {
    [models_1.SourceEcosystem.GITHUB_COPILOT]: [
        ".github/copilot-instructions.md",
        ".github/instructions",
        ".github/skills",
        ".github/agents",
        ".github/prompts",
    ],
    [models_1.SourceEcosystem.CLAUDE]: [
        "CLAUDE.md",
        ".claude/skills",
        ".claude/agents",
        ".claude/hooks",
        ".claude/settings.json",
        ".claude/rules",
    ],
};
/**
 * Join two POSIX path fragments, collapsing redundant separators.
 *
 * @param root Base POSIX path.
 * @param relative Path relative to the base.
 * @returns The combined POSIX path.
 */
function joinPosix(root, relative) {
    const normalizedRoot = root.replace(/\/+$/, "");
    const normalizedRelative = relative.replace(/^\/+/, "");
    if (normalizedRoot === "") {
        return normalizedRelative;
    }
    if (normalizedRelative === "") {
        return normalizedRoot;
    }
    return `${normalizedRoot}/${normalizedRelative}`;
}
/**
 * Resolve `.` and `..` segments in a POSIX path without touching the
 * filesystem, mirroring `Path.resolve` segment collapsing for already-absolute
 * inputs.
 *
 * @param value POSIX path that may contain `.`/`..` segments.
 * @returns The path with relative segments collapsed.
 */
function collapseSegments(value) {
    const isAbsolute = value.startsWith("/");
    const driveMatch = /^([A-Za-z]:)(.*)$/.exec(value);
    const drive = driveMatch ? (driveMatch[1] ?? "") : "";
    const body = driveMatch ? (driveMatch[2] ?? "") : value;
    const segments = body.split("/");
    const resolved = [];
    // Walk each segment, dropping "." and popping the stack for "..".
    for (const segment of segments) {
        if (segment === "" || segment === ".") {
            continue;
        }
        if (segment === "..") {
            if (resolved.length > 0) {
                resolved.pop();
            }
            continue;
        }
        resolved.push(segment);
    }
    const joined = resolved.join("/");
    if (drive) {
        return `${drive}/${joined}`;
    }
    return isAbsolute ? `/${joined}` : joined;
}
/**
 * Determine whether a candidate path is absolute (POSIX or Windows drive).
 *
 * @param value Candidate path.
 * @returns True when the path is absolute.
 */
function isAbsolutePath(value) {
    return value.startsWith("/") || /^[A-Za-z]:/.test(value);
}
/**
 * Normalize one candidate path relative to the declared source root.
 *
 * Mirrors `_normalize_relative_path`: an absolute candidate is collapsed
 * directly; a relative candidate is joined under the source root first. The
 * result must be contained within the source root or an error is raised with
 * the identical message.
 *
 * @param sourceRoot POSIX source root.
 * @param candidatePath Caller-provided source path to validate.
 * @returns The normalized source-root-relative POSIX path.
 * @throws Error When the candidate path escapes the declared source root.
 */
function normalizeRelativePath(sourceRoot, candidatePath) {
    const resolvedRoot = collapseSegments((0, file_system_1.toPosixPath)(sourceRoot));
    const candidate = (0, file_system_1.toPosixPath)(candidatePath);
    const resolvedCandidate = isAbsolutePath(candidate)
        ? collapseSegments(candidate)
        : collapseSegments(joinPosix(resolvedRoot, candidate));
    const rootPrefix = `${resolvedRoot.replace(/\/+$/, "")}/`;
    if (resolvedCandidate === resolvedRoot) {
        return "";
    }
    if (!resolvedCandidate.startsWith(rootPrefix)) {
        throw new Error(`Selected path escapes the declared source root: ${candidatePath}`);
    }
    return resolvedCandidate.slice(rootPrefix.length);
}
/**
 * Normalize selected source paths beneath the declared source root.
 *
 * Mirrors `normalize_selected_paths`: deduplicates, validates each path, and
 * returns the unique normalized relative paths sorted by POSIX text.
 *
 * @param sourceRoot POSIX source root that bounds valid source input.
 * @param selectedPaths Caller-selected paths beneath the root.
 * @returns Unique normalized relative paths sorted by POSIX text.
 * @throws Error When any selected path escapes the declared source root.
 */
function normalizeSelectedPaths(sourceRoot, selectedPaths) {
    const normalized = new Set();
    // Validate and collect every selected path; a Set deduplicates collisions.
    for (const candidatePath of selectedPaths) {
        normalized.add(normalizeRelativePath(sourceRoot, candidatePath));
    }
    return [...normalized].sort((left, right) => left < right ? -1 : left > right ? 1 : 0);
}
/**
 * Recursively collect file paths beneath one absolute directory.
 *
 * @param fileSystem Injected filesystem.
 * @param absoluteDir Absolute POSIX directory to walk.
 * @param out Accumulator receiving discovered absolute file paths.
 */
function collectFilesRecursive(fileSystem, absoluteDir, out) {
    // Walk immediate children; recurse into directories and record files.
    for (const childName of fileSystem.listDirectory(absoluteDir)) {
        const childPath = joinPosix(absoluteDir, childName);
        if (fileSystem.isDirectory(childPath)) {
            collectFilesRecursive(fileSystem, childPath, out);
        }
        else if (fileSystem.isFile(childPath)) {
            out.push(childPath);
        }
    }
}
/**
 * Yield supported artifacts for one source ecosystem.
 *
 * Mirrors `_iter_supported_artifacts`: enumerates existing supported files
 * beneath the configured ecosystem roots and returns unique source-root
 * relative paths sorted by POSIX text.
 *
 * @param fileSystem Injected filesystem.
 * @param sourceRoot POSIX source tree root.
 * @param sourceEcosystem Ecosystem defining supported surfaces.
 * @returns Unique supported relative file paths beneath the source root.
 */
function iterSupportedArtifacts(fileSystem, sourceRoot, sourceEcosystem) {
    const resolvedRoot = collapseSegments((0, file_system_1.toPosixPath)(sourceRoot));
    const rootPrefix = `${resolvedRoot.replace(/\/+$/, "")}/`;
    const discovered = new Set();
    // Enumerate only the supported ecosystem surfaces so classification starts
    // from the approved input set.
    for (const supportedRoot of SUPPORTED_ROOTS[sourceEcosystem]) {
        const absoluteSupportedRoot = collapseSegments(joinPosix(resolvedRoot, supportedRoot));
        if (!fileSystem.exists(absoluteSupportedRoot)) {
            continue;
        }
        if (fileSystem.isFile(absoluteSupportedRoot)) {
            discovered.add(absoluteSupportedRoot.slice(rootPrefix.length));
            continue;
        }
        const files = [];
        collectFilesRecursive(fileSystem, absoluteSupportedRoot, files);
        // Record each discovered file as a source-root-relative path.
        for (const filePath of files) {
            discovered.add(filePath.slice(rootPrefix.length));
        }
    }
    return [...discovered].sort((left, right) => left < right ? -1 : left > right ? 1 : 0);
}
/**
 * Discover source artifacts in deterministic normalized-relative-path order.
 *
 * Mirrors `discover_source_artifacts`: enumerates supported artifacts, then
 * filters to caller-selected files and the files contained beneath selected
 * directories when a selection is provided.
 *
 * @param fileSystem Injected filesystem.
 * @param sourceRoot POSIX root directory of the source ecosystem tree.
 * @param sourceEcosystem Declared source ecosystem.
 * @param selectedPaths Optional caller-selected subset of files or directories.
 * @returns Supported source artifact paths relative to the source root, sorted
 *   by POSIX text.
 * @throws Error When selected paths escape the source root.
 */
function discoverSourceArtifacts(fileSystem, sourceRoot, sourceEcosystem, selectedPaths) {
    const allArtifacts = iterSupportedArtifacts(fileSystem, sourceRoot, sourceEcosystem);
    if (selectedPaths === undefined) {
        return allArtifacts;
    }
    const normalizedSelected = normalizeSelectedPaths(sourceRoot, selectedPaths);
    if (normalizedSelected.length === 0) {
        return allArtifacts;
    }
    const selectedSet = new Set(normalizedSelected);
    const matched = [];
    // Include files selected directly and files contained beneath selected
    // directories so callers can target either shape deterministically.
    for (const artifactPath of allArtifacts) {
        if (selectedSet.has(artifactPath)) {
            matched.push(artifactPath);
            continue;
        }
        if (isUnderSelectedDirectory(artifactPath, selectedSet)) {
            matched.push(artifactPath);
        }
    }
    return matched.sort((left, right) => left < right ? -1 : left > right ? 1 : 0);
}
/**
 * Determine whether an artifact path lies beneath any selected directory,
 * mirroring the Python `any(parent in artifact_path.parents ...)` check.
 *
 * @param artifactPath Source-root-relative artifact path.
 * @param selectedSet Normalized selected relative paths.
 * @returns True when any selected path is an ancestor directory of the artifact.
 */
function isUnderSelectedDirectory(artifactPath, selectedSet) {
    const segments = artifactPath.split("/");
    // Build each strict ancestor prefix and test membership against the set.
    for (let index = 1; index < segments.length; index += 1) {
        const ancestor = segments.slice(0, index).join("/");
        if (selectedSet.has(ancestor)) {
            return true;
        }
    }
    return false;
}
